### ISSUE TITLE
Is it possible to output exact tiling?

### ISSUE THREAD

**Issue by chrisroat** [abria/TeraStitcher#59]
For output, either in teraconverter or paraconverter mode, is it possible (or would you accept an update to the code) to force an exact tile-size based on height/width/depth, even if there are some small edge tiles? The default seems to load balance the output into tiles that are approximately the same size, using the input height/width/depth as rough guidelines.

My use case is a secondary step to create a precomputed cloud-volume -- or any format viewable by neuroglancer. My current understanding is that the parallel uploading of such a volume would benefit from exact tiling, preferably with dimensions in powers of two.

**Comment by iannellog** [comment]
Unfortunately this is not possible at the moment.
I will evaluate the required changes and if they are minimal I will include this feature.
Leave open the issue.

**Comment by chrisroat** [comment]
Thank you.  I am happy to help as well, if you can give some small direction.

**Comment by iannellog** [comment]
The changes were quite minimal.
I have updated the repository. Now version of TeraConverter is 3.1.20.
To enable fixed (or exact) tiling you have to specify the command line option:

--fixed_tiling

which is a flag. 
With this option, all tiles have the size specified  by options --height, --width, --depth, with a possible smaller last tile. If the option is omitted, tiling is performed so that all tiles have equal size (with a difference at most 1).

**Comment by chrisroat** [comment]
Thank you so much!  We'll surely make use of these.  I'll have to vet first, that there aren't any substantial things to look out for in upgrading to & building the latest version (I'm using the portable-1.11.6-Linux download)

**Comment by iannellog** [comment]
If you can wait until monday I will generate a new 1.11.xx version for linux
The source code in the repository does neither include cuda suport  nor a few bug fixes
Unfortunately I cannot generate a new version for windows at the moment, bit for linux and macos yes

**Comment by chrisroat** [comment]
No rush at all!  (We are not yet using cuda, but bugfixes will be welcomed.)

C

**Comment by chrisroat** [comment]
Have you been able to create an new linux version?  I am still using 1.11.6.

**Comment by iannellog** [comment]
I have almost done. 
As soon as the new release is available (I hope tomorrow) for download I will inform you through this thread.

**Comment by chrisroat** [comment]
Great!  I also wanted to report that with `teraconverter`, I believe when I set `--D0=X --D0=Y` that I believe the z range I got was `[0, Y-X]` and not `[X, Y]`.  Would you be able to confirm that (and if so, add it to the release)?  I can file a separate issue when I'm back in the office, but I it's something I just encountered and won't have a reproducible test case until then.

**Comment by iannellog** [comment]
You can download the last release from

https://github.com/abria/TeraStitcher/wiki/Binary-packages

Consider that the option --fixed_tiling is available in the command -line version only and it is not exposed in the GUI. It is available both in TeraStitcher and TeraConverter. Consider also that this option cannot be used at the moment (i.e. it does not work correctly) if you use the python scripts parastitcher and paraconverter (i.e. if you want to perform image fusion in parallel).

As to the options --D0=X and --D1=Y, they set the Z range to [X, Y), i.e. slice Y is not included, but the range starts from slice X. 

Consider however that if tiles have disalignments along Z, the first slice considered in the stitched image (i.e. slice 0) is the first one that has contributions from all tiles. The same behavior for the last slice. In other words the stitched image may have less slices than tiles have before stitching. This is a design choices we made at the beginning and it cannot be changed easily.

**Comment by chrisroat** [comment]
Thank you for implementing this functionality.  My reason for needing it was explicitly because I use paraconverter scripts and output large number of tiles in parallel.  I will look into the para* scripts and see how easy it will be to update to use this new functionality.  Thanks again.

In case it wasn't clear, I was getting the range `[0, Y-X)` with `--D0=X --D1=Y`.  I will double check and provide a reproducible test case and create another issue.

**Comment by iannellog** [comment]
I have added to paraconverter the support for the option --fixed_tiling
I just did a few tests so it is possible that it does not work correctly in other cases
Since I cannot carry out more tests in the next days I add the new sript to this comment in case you want to test it as soon as possible
Please let me know if you have problems
[Paraconverter2.3.3.py.zip](https://github.com/abria/TeraStitcher/files/3908729/Paraconverter2.3.3.py.zip)

**Comment by chrisroat** [comment]
@iannellog 

I am trying the 1.11.10 terastitcher binary.  It is failing for me, and I get the message below during step 1 (import).  I do not get this error with 1.11.6.  My parameters are:
```
-1 --imin_channel=0 --volin_plugin="TiledXY|3Dseries" --imout_depth=16 --imin_plugin=IMS_HDF5
```

Error:
```
ERROR: Cannot find 2D I/O plugin "IMS_HDF5" or it is not a 2D I/O plugin
```
Is there a good way to resolve the issue?

**Comment by iannellog** [comment]
I will check, but it is likely that there is still a minor bug in this version.
I should able to fix it in a couple of days. I will inform you as soon as the problem is solved.
Just to be sure, could you please tell me:
- which platform are you using (MacOS/Linux/Windows)
- if you are using either the two-level hierarchy of folders or an externally generated xml import file for importing the dataset

**Comment by chrisroat** [comment]
We are using Linux with externally generated xml.

Thanks for taking a look!

**Comment by iannellog** [comment]
I forgot to include the IMS_HDF5 plugin when I compiled the Linux distribution.
This is why you got the error.
Download again the distribution. Please confirm me that now it works.
I apologize for the inconvenient.

**Comment by chrisroat** [comment]
I can verify that it does work to output fixed tiles.  I have another issue which I'll follow-up with separately.

**Comment by chrisroat** [comment]
Actually, I see the issue is already there about python3 with #52  -- the Paraconverter above only works in python2.  Is there a chance you could apply changes to update to python 3.7?

**Comment by chrisroat** [comment]
[FWIW, I did get the paraconverter working in python3 with [this small change](https://github.com/abria/TeraStitcher/issues/52#issuecomment-583604750)]

With fixed tiling, I found that there is an error output when reading the final tile: the bottom-right tile of the last channel. 

```
from skimage.io import imread    
fname = 'path/to/last/channel/right/bottom/corner.tif'                                                                                                           
a=imread(fname)                                                                                                                             
.../skimage/external/tifffile/tifffile.py:2618: RuntimeWarning: py_decodelzw encountered unexpected end of stream
  strip = decompress(strip)
...//skimage/external/tifffile/tifffile.py:2553: UserWarning: unpack: buffer size must be a multiple of element size
  warnings.warn("unpack: %s" % e)
```

**Comment by iannellog** [comment]
The script that can be downloaded from GitHub contains the bug fix you suggest.

I have tested paraconverter with fixed tiling and the right-bottom corner was Ok.
This is the command I issued:

mpirun -np 5 python /Users/iannello/Home/LaTeX/2015-ParTeraStitcher/paratools/last/paraconverter2_3_5.py --sfmt="TIFF (unstitched, 3D)" --dfmt="TIFF (tiled, 3D)" -s=/Users/iannello/Home/Windows/TeraTools/TestData/UCL/xml_files/xml_import_ch2L0.xml -d=/Users/iannello/Home/Windows/TeraTools/TestData/temp --depth=50 --height=512 --width=512 --fixed_tiling

Attached you will find the right-bottom tile that was generated. Please, check if is correctly opened by skimage

[030460_046590_000000.tif.zip](https://github.com/abria/TeraStitcher/files/4215474/030460_046590_000000.tif.zip)

**Comment by chrisroat** [comment]
That file does open for me without errors.  It could be that mine is multi-channel (mine was only bad on the last channel), or something to else do with the data or environment.  I can try to pare down my docker container and give a small reproducible example.
